### DISCUSSION TITLE
Centralization of PSDTools Applications

### DISCUSSION THREAD

**Discussion by ParkerH27** [PSDTools/.github#7] [Ideas]
> Let's make many apps into one app.[^1]

I am planning to start development on a website(later a PWA) that wraps our tools into one. I think we should continue separate repositories for the time being but later everything should be merged into a centralized app
My thoughts about the app are that it should be dashboard-like where you can select what you want to use. 
I am also planning on asking around the student and staff base to see if there are any ideas for helpful tools we can bundle into this app. I do not like the idea of 2 different apps (one for teachers and one for students) so it would be cool if this centralized app communicated with it's service worker to only cache the tools that a particular role would need. This would enable easy management of the app as you dont have to separately build two apps.
[^1]: App refers to a website that in the future will be PWA packaged for offline use and local installation.

**Comment by ParkerH27** [comment]
I also noticed that the way the PWA Manifest is set up, it installs the app from the https://psdtools.github.io/ url. So technically any website as long as it is set up to cache(in the service worker file) and is a repository on our gh org should work.

If this is the case, all of our apps can stay in separate repositories(will help with managing and updating them without messing up other things) and we can have one central repository that holds the PWA package and central dashboard

**Reply by lishaduck** [reply]
Using flutter doesn't prevent this. We can package it as a non-pwa, and then use one `sw` to rule them all.

**Reply by PetalCat** [reply]
We could probably build a overarching pwa that is separate from the wallet and then build all of the same tools into the wallet/flutter so if the wallet is ever decommissioned there is still a pwa that has everything besides the wallet (since the wallet requires a backend server) and while the wallet is live the flutter app has all of the other tools built in so the pwa isn't really necessary(until the wallet inevitably meets it's doom)

**Reply by PetalCat** [reply]
Cuz you know they're going to end the project at some point I just hope we get a good 5 to 10 years out of it

**Reply by PetalCat** [reply]
Hopefully we build it robustly enough for that

**Reply by lishaduck** [reply]
I just want it to be around still when I graduate.
Also, the monorepo model has many benefits. Merging them is beneficial in the long run, as code can be shared much easier, and there are fewer duplicate dependency updates.

**Reply by lishaduck** [reply]
Also, workbox (which is what powers the vite plugin we use for pwa-ing) wouldn't like having stuff outside of its control to cache.

**Reply by PetalCat** [reply]
> I just want it to be around still when I graduate. Also, the monorepo model has many benefits. Merging them is beneficial in the long run, as code can be shared much easier, and there are fewer duplicate dependency updates.

I think you have me hooked on the monorepo we just need to figure out the way we can build the pre-made tools into flutter then make the repo very clearly organized accordingly

**Reply by lishaduck** [reply]
> > I just want it to be around still when I graduate. Also, the monorepo model has many benefits. Merging them is beneficial in the long run, as code can be shared much easier, and there are fewer duplicate dependency updates.
> 
> I think you have me hooked on the monorepo we just need to figure out the way we can build the pre-made tools into flutter then make the repo very clearly organized accordingly

If we have the JS code, it should be trivial to switch it to Flutter, because Dart syntax is based on JS. Hopefully. It depends on how powerful the Standard Library is, and how Flutter architectures stuff. Calculating a GPA, though, shouldn't be difficult.

**Reply by lishaduck** [reply]
> I think you have me hooked on the monorepo we just need to figure out the way we can build the pre-made tools into flutter then make the repo very clearly organized accordingly

Good, because looking back, I realized that I already said it was a monorepo in the repo description. Though, it really is seeing as the server and UI are in the same repo rather than being distributed.

**Comment by lishaduck** [comment]
I think that a central app is a good idea. One backend, two frontends—one Flutter, one JS(or, if you're like me, Elm). The backend could be set up in Azure, as it's basically just a proxy, so the free plan would probably work. The wallet backend is just an intensive plugin, due to the modular extensibility of the backend.

**Reply by lishaduck** [reply]
Building the JS frontend for the wallet can be done later, we need to write Flutter stuff first.

**Reply by lishaduck** [reply]
Once we have a Flutter app, porting it to JS should be trivial, as should porting the JS to Flutter.
However, that's not part of the ELO, so any work of that sort should be delayed at least until September.

**Reply by lishaduck** [reply]
Flutter supports a feature called [flavors](https://docs.flutter.dev/deployment/flavors), which could be used to minimize the differences between teachers and students and remove maintenance burden, although it probably won't give us the fine-grained level of control we need, as it requires multiple apps, instead of bundling them as one. I, for one, don't like the Canvas setup with multiple duplicate apps.

**Reply by PetalCat** [reply]
> Flutter supports a feature called [flavors](https://docs.flutter.dev/deployment/flavors), which could be used to minimize the differences between teachers and students and remove maintenance burden, although it probably won't give us the fine-grained level of control we need, as it requires multiple apps, instead of bundling them as one. I, for one, don't like the Canvas setup with multiple duplicate apps.

Yess that is what I envision when I see this app... One app that can just decide hey! I want to be a teacher(aka teacher logs in) or hey! I'm a student now!(when a student email logs in)

**Reply by PetalCat** [reply]
> I think that a central app is a good idea. One backend, two frontends—one Flutter, one JS(or, if you're like me, Elm). The backend could be set up in Azure, as it's basically just a proxy, so the free plan would probably work. The wallet backend is just an intensive plugin, due to the modular extensibility of the backend.

I'm pretty sure we discussed this at lunch and now it would just be one flutter frontend that (since it is flutter) can be a website or a pwa or honestly anything it can dream up and when the server goes down the rest of the tools built in still work

**Reply by lishaduck** [reply]
> > I think that a central app is a good idea. One backend, two frontends—one Flutter, one JS(or, if you're like me, Elm). The backend could be set up in Azure, as it's basically just a proxy, so the free plan would probably work. The wallet backend is just an intensive plugin, due to the modular extensibility of the backend.
> 
> I'm pretty sure we discussed this at lunch and now it would just be one flutter frontend that (since it is flutter) can be a website or a pwa or honestly anything it can dream up and when the server goes down the rest of the tools built in still work

Correct

**Reply by lishaduck** [reply]
- <https://pub.dev/packages/flutter_tailwindcss>
- <https://pub.dev/packages/tailwind_cli>
